### PR TITLE
fix: heartbeat reliability

### PR DESCRIPTION
**Pull request by grutt**
Run heartbeat on worker thread to prevent io blocking

### CODE REVIEW

File: src/clients/dispatcher/heartbeat/heartbeat-controller.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  heartbeatWorker: Worker | undefined;
+
+  constructor(client: DispatcherClient, workerId: string) {
+    this.config = client.config;
+    this.client = client.client;
+    this.workerId = workerId;
+    this.logger = new Logger(`HeartbeatController`, this.config.log_level);
+  }
+
+  async start() {
```
Comment (abelanger5) [inline src/clients/dispatcher/heartbeat/heartbeat-controller.ts:24]: Why are these `async`?

File: src/clients/dispatcher/heartbeat/heartbeat-worker.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          );
+        }
+
+        this.logger.debug(`Heartbeat sent ${actualInterval}ms ago`);
+        this.timeLastHeartbeat = now;
+      } catch (e: any) {
+        if (e.code === Status.UNIMPLEMENTED) {
+          // break out of interval
+          this.logger.error('Heartbeat not implemented, closing heartbeat');
+          this.stop();
```
Comment (abelanger5) [inline src/clients/dispatcher/heartbeat/heartbeat-worker.ts:63]: Do we leak a thread here if we call `this.stop()` without the parent calling `terminate` on the worker?

File: src/util/thread-helper.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+export function runThreaded(scriptPath: string, options: WorkerOptions) {
+  const resolvedPath = require.resolve(scriptPath);
+
+  const isTs = /\.ts$/.test(resolvedPath);
+
+  const ex = isTs
+    ? `
+    const wk = require('worker_threads');
+    require('tsconfig-paths/register');
+    require('ts-node').register({
```
Comment (abelanger5) [inline src/util/thread-helper.ts:13]: Would this impact any environments that `ts-node` is now a runtime dependency? And does this work across node runtimes, including `bun`?

**Review by abelanger5 — APPROVED** [review]

File: src/clients/dispatcher/heartbeat/heartbeat-controller.ts (hunk 1 above)
Comment (grutt) [reply src/clients/dispatcher/heartbeat/heartbeat-controller.ts:24]: the heartbeat is async

File: src/clients/dispatcher/heartbeat/heartbeat-worker.ts (hunk 2 above)
Comment (grutt) [reply src/clients/dispatcher/heartbeat/heartbeat-worker.ts:63]: yes, but there's no way to terminate the worker thread from with in the worker thread
we should never be calling HeartbeatWorker.stop() directly, instead calling stop from the controller.

File: src/util/thread-helper.ts (hunk 3 above)
Comment (grutt) [reply src/util/thread-helper.ts:15]: this is wonky. the inline execution is only for development. the published sdk will resolve a js file which is called directly

added a note

File: src/clients/dispatcher/heartbeat/heartbeat-controller.ts (hunk 1 above)
Comment (abelanger5) [reply src/clients/dispatcher/heartbeat/heartbeat-controller.ts:24]: It doesn't `await` anything and these don't seem to be awaited, what am I missing?

Comment (grutt) [reply src/clients/dispatcher/heartbeat/heartbeat-controller.ts:24]: >     await beat();

Comment (grutt) [reply src/clients/dispatcher/heartbeat/heartbeat-controller.ts:24]: not awaited because its a top level function

**Pull request merged**